R/add_captions.R

```diff
@@ -0,0 +1,86 @@
+#' @title Touch up figure and table captions after using `convert_docx_to_rmd()`
+#'
+#' @description After using `convert_docx_to_rmd()`, figure and table captions
+#' are part of the main text and cross references to them are just text and
+#' numbers, making it difficult to integrate these `.Rmd` files in reports:
+#' they are not automatically renumbered, these captions cannot be listed,...
+#' This function replaces the figure and table descriptions with true captions,
+#' and cross references to these figures and tables with dynamic references
+#' (that refer to the figure or table label).
+#'
+#'
+#' @details This function expects an input that is generated by
+#' `convert_docx_to_rmd()`, with \itemize{
+#'   \item  figure captions below the figure and starting with `Figuur `,
+#'   then a unique number and finally the description in one paragraph
+#'   \item  table captions above the table and starting with `Tabel `,
+#'   then a unique number and finally the description in one paragraph
+#'   \item figures consisting of one image, not 2 figures near each other
+#'   \item cross references having the same label as the figure or table they
+#'   refer to, so `Figuur ` or `Tabel ` followed by the unique number written
+#'   in exactly the same way
+#' }
+#' Please check carefully for mistakes after using this function
+#'
+#' @param from The `.Rmd` file to convert.
+#' Can be given as an absolute or relative path.
+#' @param to The filename to write the resulting `.Rmd` file.
+#' Can be given as an absolute or relative path.
+#'
+#' @export
+#'
+add_captions <- function(from, to) {
+  text <- readLines(from)
+  text_1string <- paste(text, collapse = "\n")
+  # replace figure caption
+  text_1string <-
+    gsub(
+      pattern = "!\\[[^]]*\\]\\(([^\\)]*)\\)\\{([^\\}]*)\\}\\n\\n(Figuur \\d*\\D*\\d*)[.:]? (.+?)\\n\\n", #nolint
+      replacement = "![\\4 \\\\label{\\3}](\\1){\\2}\n\n", #nolint
+      x = text_1string
+    )
+  # replace figure reference
+  text_1string <-
+    gsub(
+      pattern = "([^label\\{])(Figuur \\d*\\D*\\d*)([^\\}])",
+      replacement = "\\1Figuur \\\\ref{\\2}\\3",
```

Comment:
Is `\ref{}` an rmarkdown syntax for cross-references? After `convert_from_docx()` the Rmd file is further processed in `create_protocols()` to be ready for rendering in one of the Bookdown formats. Bookdown uses `\@ref(fig:foo)` for figure environments or more generally see https://bookdown.org/yihui/rmarkdown-cookbook/cross-ref.html.

Reply:
To be sure, I tried your suggestion, but as expected, `\@ref(fig:foo)` does not change to a cross reference while rendering (at least not when building a pdf using inbomd), while `\ref{image_label}` does.
The reason: `foo` has to refer to a chunk and the images are not placed in chunks in the output of `convert_from_docx_to_rmd()`,  so `\ref{image_label}` has to be used to refer to these images.
(As I am not working on protocols, I do not need the `create_protocols()` step, I just put the resulting `.Rmd` files together in one bookdown folder and it renders fine like that, except for the fact that tables are not always presented nicely by `convert_from_docx_to_rmd()`...)

Is there a problem with the cross references in your case?  Can you show me an example of it?

Reply:
Actually, I've never seen `\ref{}` type references before and can't find documentation about it. I wonder if `\@ref(image_label)` would work as well in your example?

Reply:
No, I tried this as well after your suggestion, I'm not sure anymore if the `\` showed up in the rendered text, but furthermore it was kept literally, no conversion to a cross reference.
Honestly I also had to search for it, as I am always working with R chunks, I had never refered to an image or table in Markdown before.

Reply:
I found [this post](https://community.rstudio.com/t/cross-reference-images-in-rmarkdown/83402/2) which explains how the `\@ref(fig:foo)` syntax can be used also for `![]()` if you place `![(#fig:foo) my caption]()` in it. I would prefer that solution because it avoids confusion between the two sorts of syntaxes (i.e. stick to the one that bookdown uses). It also makes it easier when we manually change the `![]()` to a chunk and use `include_graphics()` after the conversion has been done. Would that be OK and possible?

Reply:
I tried this, but after rendering I get the following for a figure caption:
`Figuur 3.1: (#fig:Figuur 5 1) my caption` (where the original `-` between `5` and `1` apparently disappeared)
For cross references to this figure, I get:
`Figuur @ref(fig:Figuur 5 1)`

I replaced all figure captions and cross references in the whole document, as well as all table captions, but I keep having this problem.

On the other hand, when rendering a whole report, including chapters with figures in chuncks and chapters including figures without chunks, it all renders fine (except for the one chapter where I made the above changes).  So combining them is apparently not a problem, and for the purpose of composing a report with `.Rmd` and `.docx` delivered chapters, it seems just fine to use `![]()`.

Would it help if I write a function (or chunk of code) to replace what you need to be replaced afterwards?  Then I suggest you make the main workflow and add `gsub()`, and I add the regex to it.  ;-)  (To not confuse users, I suggest adding it to your workflow at the place where `![]()` is changed to chunks (I guess this could also be done with `gsub()`) and not try to add it to function `add_captions()`.)

Reply:
I think the problem why it didn't work may be the space in `Figuur 5-1`.  You should think of `foo` in `(fig:foo)` as a chunk name. A chunk name can only contain alphanumerics, no spaces, no underscores. So in the regex if you remove the spaces it should work 🤞

Reply:
I also tried replacing the space with `_` and `-`, but also this gave a similar result.  (I checked this before I replaced all figure and table captions in the whole document, but I supose the latter is not the problem, as the combination in different chapters seems to work fine.)  So no, I guess I tried more or less all possible combinations, and none of it seems to render properly using the inbomd layout.  (Honestly I only checked with the inbomd layout, as I need this layout anyway, so it could be just not supported by inbomd...)

Reply:
Ok, I checked in a clean `.Rmd` file ('html_document) with the simple code below, and it is not an  inbomd issue:
Code:
```
![(#fig:test) Alert](C:/R/gitrepos/protocolhelper/inst/css/img/alert.png)

reference to fig \@ref(fig:test)
```
Result:
```
<image>
(#fig:test) Alert

reference to fig @ref(fig:test)
```
I don't understand how you get this working.  I tried every possibility you gave, but it simply isn't rendering correctly with the chunk variant while adding a figure via a link.  And the above example is the most basic way I can possibly test it...  So unless you have a solution that truly renders at your computer, I guess we better stick to the current method: it renders properly, gives the result we want, and it is no problem at all to mix it with chunk generated images.

Reply:
OK, I think I see now what is going on: the `\ref{}` syntax is [LaTeX specific syntax](https://en.wikibooks.org/wiki/LaTeX/Labels_and_Cross-referencing). It works in your case because the output you want is a pdf. The protocols are however rendered to html by default. So, I think the `\ref{}` solution is not appropriate here. Can you try again removing the space and render to html to see if that works?

Reply:
Which space?  There is no space in the above example.

Reply:
Our replies crossed. So my answer was with regard to https://github.com/inbo/protocolhelper/pull/48#discussion_r636846228. 
In https://github.com/inbo/protocolhelper/pull/48#discussion_r636862160 did you render to `html_document`? It should be `bookdown::html_document2` to be able to use the `\@ref()` syntax.

Reply:
Eh, but I would like to render to a `bookdown::pdf_book`.  And the whole idea of Rmarkdown is that it can be rendered to all different formats.  So if we change the syntax to specific `bookdown::html_document2` syntax, it cannot be used for other documents anymore (and even not for all html documents).  Isn't there any syntax that is generic Rmarkdown that can be used for all formats?

Reply:
Or some specific html format that is not showing up in pdf, so we can put it near the LaTeX specific syntax?

Reply:
The generic way of doing it is explained [here](https://bookdown.org/yihui/rmarkdown-cookbook/cross-ref.html) and [here, scroll down to the end of that page to read about knitr::include_graphics](https://bookdown.org/yihui/bookdown/figures.html). So to have a generic solution we would need to:

- use `bookdown` 's `\@ref()` syntax (in combination with _any_ of bookdown's output formats - it is not specific to html_document2)
- use `knitr::include_graphics()` instead of `![]()`

As explained, the `\ref{}` syntax will only work for tex/pdf output.

Reply:
I tried your example and it seems to work fine:

![image](https://user-images.githubusercontent.com/18713718/119137755-7431e680-ba41-11eb-8e31-0870226ff84f.png)

Reply:
That means using `knitr::include_graphics()` is not strictly necessary when we use the hack `![(#fig:foo) caption]()`, although it is advised to do so by the author of bookdown.

Reply:
I just checked with `bookdown::pdf_document2` and that works as well.

Reply:
Maybe, but it doesn't work with `pdf_book`, which is used in inbomd...

Reply:
> * use `bookdown` 's `\@ref()` syntax (in combination with _any_ of bookdown's output formats - it is not specific to html_document2)
> * use `knitr::include_graphics()` instead of `![]()`

Ok for figures, but what do we do with tables in this case???

Reply:
Changing to `bookdown::pdf_book` works also on my laptop. 
Tried using INBOmd template, but it complained:
![image](https://user-images.githubusercontent.com/18713718/119140267-3e423180-ba44-11eb-8573-bad1a99926cb.png)


However, because `bookdown::pdf_book` works, there must be something wrong with INBOmd or with your and/or my specific installation.

Reply:
Eh, what did you try exactly?  The `![(#fig:foo) caption]`?  Does a new inbomd document render?

Otherwise you might want to install inbomd again.  (The TeX Live warning is something you can ignore, apparently it appears every year around this time.)

Reply:
Ok, `![(#fig:foo) caption]` indeed renders in a clean inbomd `pdf_book`, I should have a look what is wrong in the specific package.  And test if it renders with tables as well.

Reply:
Ok, seems to work for tables as well, so I'll change it that way?

Reply:
> Ok for figures, but what do we do with tables in this case???

From the [bookdown documentation](https://bookdown.org/yihui/bookdown/tables.html):

> You can use any types of Markdown tables in your document. To be able to cross-reference a Markdown table, it must have a labeled caption of the form `Table: (\#label) Caption here`, where `label` must have the prefix `tab:`, e.g., `tab:simple-table`.

So it seems to be the same trick.

Reply:
> Eh, what did you try exactly? The `![(#fig:foo) caption]`? Does a new inbomd document render?
````
---
title: "Untitled"
author: "Hans Van Calster"
date: "21-5-2021"
output: bookdown::pdf_book
---

# test

![(#fig:test) Alert](C:/R/GitRepositories/protocolhelper/inst/css/img/alert.png)

reference to fig \@ref(fig:test)
````

This worked as expected.

Reply:
> From the [bookdown documentation](https://bookdown.org/yihui/bookdown/tables.html):
> 
> > You can use any types of Markdown tables in your document. To be able to cross-reference a Markdown table, it must have a labeled caption of the form `Table: (\#label) Caption here`, where `label` must have the prefix `tab:`, e.g., `tab:simple-table`.
> 
> So it seems to be the same trick.

I did it this way, and it worked as well:
```
Table: (#tab:tabeltest) Een tabelhoofding

+-------------+---------------+
| test        | test          |
+-------------+---------------+

reference to tabel \@ref(tab:tabeltest)

```

Reply:
OK, after re-installing tinytex and all associated latex packages it works for INBOmd as well. See attachment.
[test_inbomd.zip](https://github.com/inbo/protocolhelper/files/6523115/test_inbomd.zip)